File: Data-Structures-and-Algorithms/0977-SquaresofaSortedArray/sortedSquares.js
const sortedSquares = function(nums) {
    var result = [];
    let l = 0;
    let r = nums.length - 1;
    while (l <= r) {
        if (Math.abs(nums[l]) < Math.abs(nums[r])) {
            result.push(nums[r]*nums[r]);
            r --;
        } else {
            result.push(nums[l]*nums[l]);
            l ++;
        }
    }
    return result.reverse()
};

nums = [-16, -7, -1, 0, 3, 7, 9]
console.log(sortedSquares(nums))

// O(n) as it iterates through the whole array (n/2 -> n)
// O(1) for space (constant extra space), if result array is not counted (which it shouldn't)
// Sort the squares of a sorted array 

File: Data-Structures-and-Algorithms/0977-SquaresofaSortedArray/sortedSquares.py
class Solution(object):
    def sortedSquares(self, nums):
        result = []
        l, r = 0, len(nums) - 1
        while l <= r:
            if abs(nums[l]) < abs(nums[r]):
                result.append(nums[r]**2)
                r -= 1
            else:
                result.append(nums[l]**2)
                l += 1
        return list(reversed(result))

nums = [-16, -7, -1, 0, 3, 7, 9]
print(Solution().sortedSquares(nums))

# O(n) as it iterates through the whole array (n/2 -> n)
# O(1) for space (constant extra space), if result array is not counted (which it shouldn't)
# Sort the squares of a sorted array 